### PR TITLE
Bump compat for Metalhead

### PR DESCRIPTION
**Pull request by ablaom**
This PR bumps the [compat] for Metalhead to "0.8" and addresses resulting breakages.

Replaces #226

### CODE REVIEW

**Comment by ablaom** [comment]
Failing on GPU only. The complaint is about scalar indexing. 

I've spent some time on this today but this is hard for me to debug because I don't currently have GPU access. I conjecture that the following code is failing on a GPU but not a CPU, and this contains the issue. Be good if someone can confirm  this indeed fails. And if so, where is the scalar indexing?

```julia
import Flux
import MLJFlux
using MLJBase
import StableRNGs.StableRNG

rng = StableRNG(123)
X, y = MLJFlux.make_images(rng);

typeof(X)
# Vector{Matrix{Gray{Float64}}}

data = MLJFlux.collate(ImageClassifier(), X, y);

Flux.gpu(data) # no effect on my CPU-only machine
typeof(data)
# Tuple{Vector{Array{Float32, 4}}, Vector{OneHotArrays.OneHotMatrix{UInt32, Vector{UInt32}}}}

n_channels = 1
n_classes = 3
init = Flux.glorot_uniform(rng)

chain = Flux.Chain(
    Flux.Conv((2, 2), n_channels=>2, init=init),
    Flux.Conv((2, 2), 2=>1, init=init),
    x->reshape(x, :, size(x)[end]),
    Flux.Dense(16, n_classes, init=init))

x = data[1][1]
typeof(x)
# Array{Float32, 4}

sizeof(x)
# (6, 6, 1, 1)

chain(x)
```

**Comment by ablaom** [comment]
Okay. I guess the `reshape` is the issue. https://github.com/JuliaGPU/CUDA.jl/issues/228

**Comment by ablaom** [comment]
Okay, looks like the colon is not supported in the `CuArray`-specialized `reshape`.

**Comment by mohamed82008** [comment]
@ablaom would you also be open to making Metalhead an optional dependency?

File: test/image.jl [hunk 1]
```diff
@@ -8,17 +8,29 @@ mutable struct MyNeuralNetwork <: MLJFlux.Builder
     kernel2
 end
 
-function MLJFlux.build(model::MyNeuralNetwork, rng, ip, op, n_channels)
+# to get a matrix whose last dimension mathces that of the array input (the batch size):
+function make2d(x)
+    l = length(x)
+    b = size(x)[end]
+    reshape(x, div(l, b), b)
```
Comment (ablaom) [inline test/image.jl:15]: So I thought this use of `reshape` would avoid the "scalar indexing error", but tests are still not passing.

**Comment by ablaom** [comment]
> @ablaom would you also be open to making Metalhead an optional dependency?

That's tricky because the default builder for ImageClassifier is a VGG architecture from Metalhead.jl. We could throw an error if `Metalhead.jl` is not loaded, but that would violate a general principle currently holding true among 200+ MLJ models (and assumed in applications of MLJTestIntegration.jl): *the empty argument constructor always works* unless the model is a wrapper, like `TunedModel`. 

An alternative, which already looks too complicated to me, is to make the default

- something simple (and essentially useless) in the case Metalhead.jl is not loaded, but throw an informative warning suggesting it be loaded.
- the VGG model otherwise

Related discussion: https://github.com/FluxML/MLJFlux.jl/issues/162

Maybe you have a  better idea?

**Comment by mohamed82008** [comment]
Not a better idea but moving ImageClassifier itself out to another package is another option. Not a great option though if you want all Flux-related wrappers to be in this repo.

**Comment by ToucheSir** [comment]
> Okay, looks like the colon is not supported in the `CuArray`-specialized `reshape`.

That seems strange to me, because it's basically what [`MLUtils.flatten`](https://github.com/JuliaML/MLUtils.jl/blob/03f8fb12420ad18766389297ca3522f52c32e46c/src/utils.jl#L489) does and I don't recall having any issues with it. Indeed, making the same reshape calls manually results in a `CuArray`.

I had a look through the failing CI runs, and the problem is instead this warning: https://buildkite.com/julialang/mljflux-dot-jl/builds/339#018a25d7-e2c7-4c2b-a7af-c1a9d97436c4/425-777. Because we switched to package extensions in Flux 0.14, cuDNN needs to be separately added to an environment to enable the CUDA conv routines in NNlib. I'm guessing MLJFlux doesn't want to take it on as a dep, so adding it into your test env/extras should be enough.

**Comment by ablaom** [comment]
Thanks indeed for the help @ToucheSir. I guess that scalar indexing error was a Red Herring. I've added `cuDNN` to the tests deps and there's no sign of it now.

**Pull request merged**